02-peer-protocol.md

```diff
@@ -1106,6 +1118,75 @@ it's simplest to only allow it to set fee levels; however, as the same
+
+To facilitate privacy of unannounced channels, peers can request a random
+short_channel_id to use which does not have any relationship to the actual
+channel funding transaction.  Note that these messages are all even, so it
+is illegal to send them to peers which do not offer `option_scid_assign`.
+
+1. type: 260 (`assign_scid`) (`option_scid_assign`)
+2. data:
+   * [`channel_id`:`channel_id`]
+
+1. type: 268 (`assign_scid_reply`) (`option_scid_assign`)
+   * [`channel_id`:`channel_id`]
+   * [`short_channel_id`:`short_channel_id`]
+
+1. type: 270 (`unassign_scid`) (`option_scid_assign`)
+2. data:
+   * [`channel_id`:`channel_id`]
+
+1. type: 272 (`unassign_scid_reply`) (`option_scid_assign`)
+   * [`channel_id`:`channel_id`]
+
+#### Requirements
+
+A sending node:
+  - if it did not offer `option_scid_assign`:
+    - MUST NOT send `assign_scid` or `unassign_scid`
+  - SHOULD remember the last `assign_scid_reply` `short_channel_id` for use in
+    receiving payments.
+  - MAY send `unassign_scid` when no invoices are pending, and send `assign_scid` when
+    it wants to create a new invoice and none are pending.
+
+A receiving node:
+  - on receiving `assign_scid`:
+    - MUST assign an unpredictable unique 64-bit number to this channel
+      (see [Assigned `short_channel_id` Forwarding](04-onion-routing.md#assigned-short_channel_id-forwarding)).
+	  and return it in an ``assign_scid_reply` message.
+	  - SHOULD ensure this does not conflict with future `short_channel_id`s.
+    - MUST discard any previous assigned `short_channel_id` for this channel
+  - on receiving `unassign_scid`:
+    - MUST discard any previous assigned `short_channel_id` for this channel
+	  - Note: it's not an error if there are no previous assignments.
+	- MUST reply with `unassign_scid_reply`.
+
+#### Rationale
+
+There are two ways of hiding a private channel's funding-transaction-derived
+`short_channel_id`.  The most thorough is `disable_incoming` which ensures
+that the private `short_channel_id` is never used for forwarding.  However,
+since channels may have been opened before `option_scid_assign` was
+implemented, assigning or unassigning a random short_channel_id has the same
+effect of disabling the original `short_channel_id`.  See [Assigned
+`short_channel_id` Forwarding](04-onion-routing.md#assigned-short_channel_id-forwarding).
+
+The space of short_channel_id is only 64 bits, so it's possible for an
+adversary to probe for them over time, using payment attempts.  Thus it's
+recommended that they only be assigned when they're actually likely to be
+used, and unassigned afterwards.
+
+If it were to conflict with a future `short_channel_id` for a real
+channel, misrouting may occur.  The simplest way to minimize this is
+to use old block numbers, but that also reduces the search space.
+Simpler is to select a random value which isn't already used, and
+close (before funding_locked) any future channel unfortunate enough to
+clash.
```

Comment:
I'm more in favor of restricting to old blocks, this makes it a lot simpler to manage.
Restricting to old blocks means we have ~500.000*2^40 possibilities ~= 10^17
Assuming 100.000 attempts per second (most nodes probably won't be able to forward that many payment attempts per second, so it feels conservative) it will take more than 150.000 years to probe.
Should be a big enough search space, isn't it? ;)

Reply:
In practice, you can just pick a random 64 you will never clash. I don't want to overly restrict implementations: you can choose to use a subset of believable unspent outputs, etc.

Reply:
Agreed, I just think we can remove the statement that choosing old blocks reduces the search space because in practice it's irrelevant.

Reply:
> Should random `scid` be the same for both peers? I suppose there's no reason for it to be different from each side as that would be easier to apply to current implementations where same `scid` is already assumed.

No, since it might clash for one side, and one side might tear it down while you were trying to use it.  And why make it the same?  That just leaks information.

Reply:
> And why make it the same? That just leaks information.

Makes implementation easier at least in Eclair where I'm trying to do it: currently graph engine determines that two `ChannelUpdate`s belong to the same channel by them having a same `scid`. But of course this should not dictate the spec.

Just curious: what kind of info is specifically leaked by same `scid`?

Reply:
The leak us:

1. Alice has a private channel with Bob: they call it scid-tmp.
2. Mallory asks Alice's website for an invoice, Alice gives Mallory a fake pubkeyA and scid-tmp from Bob.
3. Mallory asks Bob's website for an invoice, Bob gives a fake pubkeyB and the same scid-tmp from Alice.

Mallory has learned that Alice == pubkeyA and Bob == pubkeyB.